sqs/src/main/scala/akka/stream/alpakka/sqs/impl/BalancingMapAsyncUnordered.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.sqs.impl
+
+import akka.annotation.InternalApi
+import akka.stream.ActorAttributes.SupervisionStrategy
+import akka.stream.Attributes.name
+import akka.stream.impl.{ReactiveStreamsCompliance, Buffer => BufferImpl}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import akka.stream._
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success, Try}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final case class BalancingMapAsyncUnordered[In, Out](
+    initialParallelism: Int,
+    f: In ⇒ Future[Out],
+    balancingF: Out ⇒ Int
+) extends GraphStage[FlowShape[In, Out]] {
+
+  private val in = Inlet[In]("BalancingMapAsyncUnordered.in")
+  private val out = Outlet[Out]("BalancingMapAsyncUnordered.out")
+
+  override def initialAttributes = name("BalancingMapAsyncUnordered")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      override def toString = s"BalancingMapAsyncUnordered.Logic(inFlight=$inFlight, buffer=$buffer)"
+
+      val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+
+      private var inFlight = 0
+      private var buffer: BufferImpl[Out] = _
+      private var parallelism = initialParallelism
```

Comment:
Great addition to the `mapAsyncUnordered`. I think we should drop `initialParallelism` from the arguments of this stage, as whatever it is, it is reset after first Future completes. So we could have it hardcoded to 1 and still have the same bahaviour.

Reply:
yeah, slow start might make sense.

After some thought I also want to change the signature of balancing function to (Out, Int) => Int, where the second argument is the current value of parallelism, wdyt? So one can increase parallelism gradually, which might be useful for something like a circuit-breaker case.

Reply:
however I still need maxParallelism parameter to initialise the buffer inside the stage

Reply:
Balancing function that takes previous parallelism sounds good!

> however I still need maxParallelism parameter to initialise the buffer inside the stage

Ahh, thats right. What is more, the balancing function should not return larger parallelism than `maxParallelism`. But since this is not user facing API, it can be left like this for now.

Reply:
> balancing function should not return larger parallelism than `maxParallelism`
Yeah, I also thought about it

I'll create a PR to akka-streams later this week and will address this point there.

Reply:
Sounds good. Create an issue first, with your proposal, so any interested parties could have a discussion before the code gets written. :)